File: scratch/gopurs-adt-reuse-validation-20260916/before-pure/src/Test/ChurchFFI.go
package Test_ChurchFFI


type Church func(func(int) int) func(int) int

func zeroC() Church {
	return func(f func(int) int) func(int) int {
		return func(x int) int {
			return x
		}
	}
}

func succC(n Church) Church {
	return func(f func(int) int) func(int) int {
		return func(x int) int {
			return f(n(f)(x))
		}
	}
}

func addC(m Church, n Church) Church {
	return func(f func(int) int) func(int) int {
		return func(x int) int {
			return m(f)(n(f)(x))
		}
	}
}

func mulC(m Church, n Church) Church {
	return func(f func(int) int) func(int) int {
		return func(x int) int {
			return m(n(f))(x)
		}
	}
}

func fromInt(n int) Church {
	if n == 0 {
		return zeroC()
	}
	return succC(fromInt(n - 1))
}

func toInt(n Church) int {
	return n(func(x int) int { return x + 1 })(0)
}

func c10(n int) Church {
	return fromInt(n)
}

func c100(n int) Church {
	return mulC(c10(n), c10(n))
}

func c10k(n int) Church {
	return mulC(c100(n), c100(n))
}

func c100k(n int) Church {
	return mulC(c10k(n), c10(n))
}

func RunChurchFFI(limit int) int {
	dummy := limit
	return (toInt(c100k(dummy)))
}

File: scratch/record-scalarization-integration-20260916/step1/src/Test/ChurchFFI.php
<?php
$exports['runChurchFFI'] = function($limit) {
    $zeroC = function($f) { return function($x) { return $x; }; };
    $succC = function($n) {
        return function($f) use ($n) {
            return function($x) use ($n, $f) {
                return $f($n($f)($x));
            };
        };
    };
    $mulC = function($m) {
        return function($n) use ($m) {
            return function($f) use ($m, $n) {
                return function($x) use ($m, $n, $f) {
                    return $m($n($f))($x);
                };
            };
        };
    };
    $fromInt = function($n) use (&$fromInt, $zeroC, $succC) {
        if ($n === 0) return $zeroC;
        return $succC($fromInt($n - 1));
    };
    
    $c10 = function($n) use ($fromInt) { return $fromInt($n); };
    $c100 = function($n) use ($c10, $mulC) { return $mulC($c10($n))($c10($n)); };
    $c10k = function($n) use ($c100, $mulC) { return $mulC($c100($n))($c100($n)); };
    $c100k = function($n) use ($c10k, $c10, $mulC) { return $mulC($c10k($n))($c10($n)); };
    
    $toInt = function($n) {
        return $n(function($x) { return $x + 1; })(0);
    };
    
    return $toInt($c100k((int)$limit));
};
return $exports;
